README.md
#tidal-midi [![Build Status](https://travis-ci.org/tidalcycles/tidal-midi.svg?branch=master)](https://travis-ci.org/tidalcycles/tidal-midi)

A [TidalCycles](http://tidalcycles.org) module for sending patterns over MIDI.

__PortMIDI__ variant. Should work on OS X, Linux and Windows.

This _still_ is __experimental__ software.

<ul>
<li><a href="#installation">Installation</a>
  <ul>
    <li><a href="#prerequisites">Prerequisites</a></li>
    <li><a href="#install">Install tidal-midi</a></li>
  </ul>
</li>
<li><a href="#usage">Usage</a>
  <ul>
  <li><a href="#mididevices">MIDI devices on your system</a></li>
  <li><a href="#boot">Boot tidal-midi</a></li>
  <li><a href="#playingpatterns">Playing Patterns</a>
    <ul>
      <li><a href="#veldur">Note length, velocity, and other MIDI CC parameters</li>
    </ul>
  </li>
  <li><a href="#custommidichannels">Custom MIDI Channels</a></li>
  <li><a href="#multichannel">Multiple MIDI Channels</a></li>
  <li><a href="#defaultsynthcontroller">The default synthController</a></li>
  </ul>
</li>
<li><a href="#supportedsynths">Supported Synthesizers</a></li>
<li><a href="#custommappings">How to write your own synth mapping</a></li>
<li><a href="#emacs">Automatic bootup in Emacs</a></li>
<li><a href="#known_issues">Known Issues</a></li>
</ul>

<a name="installation"></a>
# Installation

<a name="prerequisites"></a>
## Prerequisites

Depending on your operating system, you will need to install some prerequisites
first.

### All Systems

`tidal-midi` requires the latest version of `tidal`. Run these two commands
in your terminal to install the latest version:

```shell
cabal update
cabal install tidal
```

### Linux

Run the following to install `libasound2-dev` and `libportmidi-dev`:

```shell
apt-get install libasound2-dev and libportmidi-dev
```

### Mac OS X

Install PortMIDI:

```shell
brew install portmidi
```

<a name="install"></a>
## Install tidal-midi

Simply do:

```shell
cabal update
cabal install tidal-midi
```

__Note:__ On OS X with GHC 7.10 it is necessary to reinstall PortMidi again with
frameworks correctly linked:

```shell
cabal install portmidi --ghc-options="-optl-Wl,-framework,CoreMIDI,-framework,CoreAudio" --reinstall --jobs=1 --force-reinstalls
```

<a name="usage"></a>
# Usage

_This guide assumes you are already familiar with Tidal and creating patterns
with samples._

<a name="mididevices"></a>
## Get the names of MIDI devices on your system

In order to use `tidal-midi` you will need the _exact_ name of a MIDI
device on your system. You can get a list of MIDI devices on your system
by running some code in a regular `.tidal` file.

Assuming you're using the Atom editor, create a new file and save it with
a `.tidal` extension (e.g. `midi-test.tidal`). Then, type the following in
the editor:

```haskell
import Sound.Tidal.MIDI.Context

displayOutputDevices >>= putStrLn
```

Evalulate both of those above lines separately using `Shift+Enter` in Atom.
After evaluating the last line, it will output a list of MIDI devices
in your editor (in Atom, at the bottom output panel).

After listing MIDI devices on your system, take note of the device name you
will use. Devices names are case-sensitive.

For the purposes of this guide, we'll assume your device name is "USB MIDI Device".

> You only need to do this step whenever you want to get a list of devices.
> Once you take note of your system's device names, you don't need to perform
> this step ever again (unless you acquire a new MIDI device).

<a name="boot"></a>
## Boot tidal-midi

Make sure you're currently working in a file with a `.tidal` extension in
your editor (it could be the same file from the device list step above).
Then type these three lines of bootup code:

```haskell
import Sound.Tidal.MIDI.Context

devices <- midiDevices

m1 <- midiStream devices "USB MIDI Device" 1 synthController
```

Evaluate each of those lines (use `Shift+Enter` in the Atom
editor). Now Atom is ready to run MIDI patterns using `m1`.

> In the last line of the boot code above, the last three parameters
> are the most important:
>
> - "USB MIDI Device" is the name of your device
> - 1 is the MIDI channel number
> - synthController is the type of synthesizer code to use (you can use custom ones)

<a name="playingpatterns"></a>
## Playing patterns on your device

The following code will play a very simple pattern on middle-C:

```haskell
m1 $ note "0"
```

Above, the `note` param indicates a MIDI note, where `0` equals middle-C. The
following pattern plays a major scale:

```haskell
m1 $ note "0 2 4 5 7 9 11 12"
```

Alternatively, you can use `midinote` to explicitly use a MIDI note from 0 to 127:

```haskell
m1 $ midinote "60 62 64 65 67 69 71 72"
```

You can use normal TidalCycles pattern transform functions to change `tidal-midi`
patterns:

```haskell
m1 $ every 3 (rev) $ every 2 (density 2) $ note "0 2 4 5 7 9 11 12"
```

<a name="veldur"></a>
### Note length, velocity, and other MIDI CC parameters

Note length and velocity are controlled using the `dur` and `velocity`
parameters, respectively.

The value of `dur` is given in seconds:

```haskell
m1 $ note "0 2" # dur "0.05 0.2"

m1 $ note "0 2" # dur (scale 0.05 0.3 $ slow 1.5 tri1)
```

Alternatively, the `legato` parameter tells Tidal to scale the note
duration to fill it's "slot" in the pattern.  For example, the following
will give four notes each a quarter cycle in duration (values of legato
  greater or less than one will multiply the duration):

```haskell
m1 $ note "0 1 0 2" # legato "1"
```

`velocity` has a range from *0 to 1*, and equates to MIDI values *0 to 127*:

```haskell
m1 $ note "0 2 4 5 7 9 11 12" # velocity "0.5 0.75 1"

m1 $ note "0 2 4 5 7 9 11 12" # velocity (scale 0.5 1 $ slow 1.5 saw1)
```

The `synthController` has some params that support MIDI Change Control messages,
such as the mod wheel:

```haskell
m1 $ note "0 2 4 5 7 9 11 12" # modwheel "0.1 0.4 0.9"
```

Details about the default MIDI CC messages can be found in the
[default synth controller](#defaultsynthcontroller) section below.

MIDI CC params can have decimal values in the range *0 to 1*, which map to MIDI
CC values *0 to 127*.

_Custom synthesizer implementations may implement additional MIDI CC parameters.
Please refer to the [supported synths](doc/synths.md) for more information._

<a name="custommidichannels"></a>
## Custom MIDI Channels

Let's review this line from the boilerplate code above:

```haskell
m1 <- midiStream devices "USB MIDI Device" 1 synthController
```

The 2nd to last parameter on that line indicates the channel number. Let's say
your device is running on channel 7. You can specify channel 7 by changing the
2nd to last parameter:

```haskell
m1 <- midiStream devices "USB MIDI Device" 7 synthController
```

<a name="multichannel"></a>
## Multiple MIDI Channels

`tidal-midi` supports devices with multiple channels so that you can create
patterns on each channel separately:

```haskell
m1 <- midiStream devices "USB MIDI Device" 1 synthController
m2 <- midiStream devices "USB MIDI Device" 2 synthController
m5 <- midiStream devices "USB MIDI Device" 5 synthController

m1 $ note (run 4) # velocity "0.5"
m2 $ note "0*2 5 7" # dur "0.1"
m5 $ midinote "36 60"
```

> Note: at the time of this writing, multiple channels can cause scheduling
> problems if the synth controller's latency is too low. This is mainly an
> issue for `tidal-midi` developers to improve, but users may be impacted.
> Latency values can be increased by modifying the synth controller's source
> code (e.g. in `SimpleSynth.hs`), then re-compiling `tidal-midi`
> with `cabal install`.

<a name="defaultsynthcontroller"></a>
## The default synthController (a.k.a "simple synth")

The simple synth comes with _simple_ MIDI parameters, that any device should understand:

* modwheel (MIDI CC #1)
* balance (MIDI CC #8)
* expression (MIDI CC #11)
* sustainpedal (MIDI CC #64)

All of these parameters map the given values from __0..1__ to MIDI values ranging from __0..127__.

You can use all of these parameters like the familiar synth parameters in
TidalCycles:

```haskell
m1 $ note "0*8" # modwheel "0.25 0.75" # balance "0.1 0.9" # expression (sine1)
```

__All tidal-midi synthesizers__ "inherit" from the "simple synth" and
automatically expose these same parameters.

See the section below on [Supported Synthesizers](#supportedsynths) for details
on custom controllers for popular hardware synthesizers.

<a name="supportedsynths"></a>
# Supported Synthesizers

A variety of custom mappings have been created in `tidal-midi` for popular hardware synthesizers.
Click on a device below to get details on its usage:

* [DSI Tetra](doc/synths.md#dsi-tetra)
* Elektron Analog RYTM
* [Korg Volca Bass](doc/synths.md#korg-volca-bass)
* [Korg Volca Beats](doc/synths.md#korg-volca-beats)
* [Korg Volca Keys](doc/synths.md#korg-volca-keys)
* Roland System-1M
* Synthino
* [Waldorf Blofeld](doc/synths.md#waldorf-blofeld)


<a name="custommappings"></a>
# How to write your own synth mapping

Interested in using `tidal-midi` with your own synthesizer? Please read the guide on [Writing a new synth mapping](doc/synth-mapping.md).


<a name="emacs"></a>
# Automatic startup in Emacs

Within your `tidal.el` script, locate the function `tidal-start-haskell` and add:

```emacs
(tidal-send-string "import Sound.Tidal.MIDI.Context")
```

after

```emacs
(tidal-send-string "import Sound.Tidal.Context")
```

Additionally you will have to add lines to import the synth you want to control via MIDI, e.g. `(tidal-send-string "import Sound.Tidal.VolcaKeys")` as well as the initialization commands for streams:

```emacs
(tidal-send-string "devices <- midiDevices")
(tidal-send-string "t1 <- midiStream devices \"USB MIDI Device\" 1 synthController")
```

The above code adds the MIDI device "USB MIDI Device" and controls it via MIDI channel 1.
With this set up you will be able to use it via e.g. `t1 $ note "50"`


<a name="known_issues"></a>
# Known issues and limitations

- SysEx support is there but really limited to work with the Waldorf Blofeld


CHANGELOG.md
# Changelog

## 1.0
- add support for `legato` parameter to automatically scale note durations

## 0.9

- normalize Shape names as lowercased Controller Module name, e.g.
  if you `import Sound.Tidal.MIDI.VolcaKeys` use `midiStream "Device Name" 1 volcakeys`
- default latency of `0.1` to work around late messages _most of the time_
- drop default values in `ControlChange`, can be specified via Tidal `Param` helpers
  `pF`, `pS` and `pI`
- default `n` to `128` to allow CC-only patterns, e.g. `m1 $ resonance "0.3 0.1 0.9"`
- params that _leave_ patterns will be reset to their defaults
- `unit "cycle"` calculates inter-onsets and tries to behave like SuperDirt
- `Stream` is now just the Tidal communication layer and internal MIDI
  messaging/scheduling lives in `Output`


Sound/Tidal/Tetra.hs
module Sound.Tidal.Tetra where

import Sound.Tidal.Context hiding (latency)
import Sound.Tidal.MIDI.Control

-- import Sound.Tidal.Parse
-- import Sound.Tidal.Pattern

-- import Control.Applicative

makeSeqTracks = concat $ map makeSeqTrack [0..3]
  where makeSeqTrack t = map (makeSeqStep t) [0..15]
        makeSeqStep t s = NRPN (snd $ makeName t s) (makeCC t s) (0, 127) 0 passThru
        makeName t s = pF ("seq" ++ (show $ t + 1) ++ "step" ++ (show $ s + 1)) (Just 0)
        makeCC t s = (t * 16) + (s + 120)

tetra :: ControllerShape
tetra = ControllerShape {
  controls = ([
                 NRPN osc1freq_p 0 (0, 120) 0.2 passThru,
                 mrNRPN osc1detune_p 1 (0, 100) 0.5,
                 NRPN osc1shape_p 2 (0, 103) 1 passThru,
                 mNRPN osc1glide_p 3,
                 NRPN osc1kbd_p 4 (0, 1) 1 passThru,

                 NRPN osc2freq_p 5 (0, 120) 0.2 passThru,
                 mrNRPN osc2detune_p 6 (0, 100) 0.5,
                 NRPN osc2shape_p 7 (0, 103) 1 passThru,
                 mNRPN osc2glide_p 8,
                 NRPN osc2kbd_p 9 (0, 1) 1 passThru,

                 NRPN oscsync_p 10 (0, 1) 0 passThru,
                 NRPN glidemode_p 11 (0, 3) 0 passThru,
                 NRPN oscslop_p 12 (0, 5) 0 passThru,
                 mrNRPN oscmix_p 13 (0, 127) 0.5,

                 mNRPN noise_p 14,
                 mrNRPN cutoff_p 15 (0, 164) 1,
                 mNRPN resonance_p 16,
                 mNRPN kamt_p 17,
                 mNRPN audiomod_p 18,
                 NRPN fpoles_p 19 (0, 1) 1 passThru,

                 -- filter envelope
                 mrNRPN famt_p 20 (0, 254) 0.5,
                 mNRPN fvel_p 21,
                 mNRPN fdel_p 22,
                 mNRPN fatk_p 23,
                 mNRPN fdcy_p 24,
                 mNRPN fsus_p 25,
                 mNRPN frel_p 26,


                 mNRPN vcavol_p 27,
                 mNRPN outspread_p 28,
                 mrNRPN gain_p 29 (0, 127) 1, -- max volume by default

                 mrNRPN vamt_p 30 (0, 127) 1, -- max vca envelope amount
                 mNRPN vvel_p 31,
                 mNRPN vdel_p 32,
                 mNRPN attack_p 33,
                 mrNRPN decay_p 34 (0, 127) 0.5,
                 mrNRPN sustain_p 35 (0, 127) 0.5,
                 mNRPN release_p 36,

                 NRPN lfo1rate_p 37 (0, 166) 0 passThru, -- unsynced
                 -- mrNRPN "lfo1step" 37 (151, 166) 0, -- (bpm / 32) - 16 bpm-cycles
                 NRPN lfo1shape_p 38 (0, 4) 0 passThru,
                 mNRPN lfo1amt_p 39,
                 NRPN lfo1dest_p 40 (0, 43) 0 passThru,
                 NRPN lfo1sync_p 41 (0, 1) 0 passThru,

                 NRPN lfo2rate_p 42 (0, 166) 0 passThru, -- unsynced
                 -- mrNRPN "lfo2step" 42 (151, 166) 0, -- (bpm / 32) - 16 bpm-cycles
                 NRPN lfo2shape_p 43 (0, 4) 0 passThru,
                 mNRPN lfo2amt_p 44,
                 NRPN lfo2dest_p 45 (0, 43) 0 passThru,
                 NRPN lfo2sync_p 46 (0, 1) 0 passThru,

                 NRPN lfo3rate_p 47 (0, 166) 0 passThru, -- unsynced
                 -- mrNRPN "lfo3step" 47 (151, 166) 0, -- (bpm / 32) - 16 bpm-cycles
                 NRPN lfo3shape_p 48 (0, 4) 0 passThru,
                 mNRPN lfo3amt_p 49,
                 NRPN lfo3dest_p 50 (0, 43) 0 passThru,
                 NRPN lfo3sync_p 51 (0, 1) 0 passThru,

                 NRPN lfo4rate_p 52 (0, 166) 0 passThru, -- unsynced
                 -- mrNRPN "lfo4step" 52 (151, 166) 0, -- (bpm / 32) - 16 bpm-cycles
                 NRPN lfo4shape_p 53 (0, 4) 0 passThru,
                 mNRPN lfo4amt_p 54,
                 NRPN lfo4dest_p 55 (0, 43) 0 passThru,
                 NRPN lfo4sync_p 56 (0, 1) 0 passThru,

                 NRPN emod_p 57 (0, 43) 0 passThru,
                 mrNRPN eamt_p 58 (0, 254) 0.5,
                 mNRPN evel_p 59,
                 mNRPN edel_p 60,
                 mNRPN eatk_p 61,
                 mNRPN edcy_p 62,
                 mNRPN esus_p 63,
                 mNRPN erel_p 64,

                 NRPN mod1src_p 65 (0, 20) 0 passThru,
                 mrNRPN mod1amt_p 66 (0, 254) 0.5,
                 NRPN mod1dst_p 67 (0, 47) 0 passThru,

                 NRPN mod2src_p 68 (0, 20) 0 passThru,
                 mrNRPN mod2amt_p 69 (0, 254) 0.5,
                 NRPN mod2dst_p 70 (0, 47) 0 passThru,

                 NRPN mod3src_p 71 (0, 20) 0 passThru,
                 mrNRPN mod3amt_p 72 (0, 254) 0.5,
                 NRPN mod3dst_p 73 (0, 47) 0 passThru,

                 NRPN mod4src_p 74 (0, 20) 0 passThru,
                 mrNRPN mod4amt_p 75 (0, 254) 0.5,
                 NRPN mod4dst_p 76 (0, 47) 0 passThru,

                 NRPN seq1dst_p 77 (0, 47) 0 passThru,
                 NRPN seq2dst_p 78 (0, 47) 0 passThru,
                 NRPN seq3dst_p 79 (0, 47) 0 passThru,
                 NRPN seq4dst_p 80 (0, 47) 0 passThru,

                 mrNRPN mwhl_p 81 (0, 254) 0.5,
                 NRPN mwhldst_p 82 (0, 47) 0 passThru,

                 mrNRPN aftt_p 83 (0, 254) 0.5,
                 NRPN afttdst_p 84 (0, 47) 0 passThru,

                 mrNRPN breath_p 85 (0, 254) 0.5,
                 NRPN breathdst_p 86 (0, 47) 0 passThru,

                 mrNRPN mvel_p 87 (0, 254) 0.5,
                 NRPN mveldst_p 88 (0, 47) 0 passThru,

                 mrNRPN foot_p 89 (0, 254) 0.5,
                 NRPN footdst_p 90 (0, 47) 0 passThru,

                 NRPN kbpm_p 91 (30, 250) 0 passThru,
                 NRPN clockdiv_p 92 (0, 12) 0 passThru, -- TODO: document values

                 NRPN bendrng_p 93 (0, 12) 0 passThru,

                 NRPN sqntrig_p 94 (0, 4) 0 passThru, -- TODO: document values
                 NRPN unisonkey_p 95 (0, 5) 0 passThru, -- TODO: document values
                 NRPN unisonmode_p 96 (0, 4) 0 passThru, -- TODO: document values
                 NRPN arpmode_p 97 (0, 14) 0 passThru,

                 NRPN erepeat_p 98 (0, 1) 0 passThru,
                 NRPN unison_p 99 (0, 1) 0 passThru,


                 NRPN arp_p 100 (0, 1) 0 passThru,
                 NRPN sqn_p 101 (0, 1) 0 passThru,

                 NRPN mcr1_p 105 (0, 183) 0 passThru,
                 NRPN mcr2_p 106 (0, 183) 0 passThru,
                 NRPN mcr3_p 107 (0, 183) 0 passThru,
                 NRPN mcr4_p 108 (0, 183) 0 passThru,

                 mNRPN fbgain_p 110,

                 mrNRPN btnfreq_p 111 (0, 127) 0.25,
                 mrNRPN btnvel_p 112 (0, 127) 1,
                 NRPN btnmode_p 113 (0, 1) 0 passThru,

                 mNRPN sub1vol_p 114,
                 mNRPN sub2vol_p 115,

                 mNRPN fbvol_p 116,

                 -- left out: editor byte,

                 mrNRPN ksplitpoint_p 118 (0, 127) 0.5,
                 NRPN kmode_p 119 (0, 2) 0 passThru, -- TODO: document values
                 mCC mute_p 123,
                 mCC ccreset_p 121,
                 mCC damp_p 64

              ] ++ makeSeqTracks),
                         latency = 0.01
                       }

tetraShape = toShape tetra


(ccreset, ccreset_p)       = pF "ccreset" (Just 0)
(damp, damp_p)             = pF "damp" (Just 0)


(osc1freq, osc1freq_p)     = pF "osc1freq" (Just 0)
(osc1detune, osc1detune_p) = pF "osc1detune" (Just 0)
(osc1shape, osc1shape_p)   = pF "osc1shape" (Just 0)
(osc1glide, osc1glide_p)   = pF "osc1glide" (Just 0)
(osc1kbd, osc1kbd_p)       = pF "osc1kbd" (Just 0)

(osc2freq, osc2freq_p)     = pF "osc2freq" (Just 0)
(osc2detune, osc2detune_p) = pF "osc2detune" (Just 0)
(osc2shape, osc2shape_p)   = pF "osc2shape" (Just 0)
(osc2glide, osc2glide_p)   = pF "osc2glide" (Just 0)
(osc2kbd, osc2kbd_p)       = pF "osc2kbd" (Just 0)

(oscsync, oscsync_p)       = pF "oscsync" (Just 0)
(glidemode, glidemode_p)   = pF "glidemode" (Just 0)
(oscslop, oscslop_p)       = pF "oscslop" (Just 0)
(oscmix, oscmix_p)         = pF "oscmix" (Just 0)

(noise, noise_p)           = pF "noise" (Just 0)
(kamt, kamt_p)             = pF "kamt" (Just 0)
(audiomod, audiomod_p)     = pF "audiomod" (Just 0)
(fpoles, fpoles_p)         = pF "fpoles" (Just 0)
twopole                    = fpoles (p "0")
fourpole                   = fpoles (p "1")

-- filter envelope
(famt, famt_p)             = pF "famt" (Just 0)
(fvel, fvel_p)             = pF "fvel" (Just 0)
(fdel, fdel_p)             = pF "fdel" (Just 0)
(fatk, fatk_p)             = pF "fatk" (Just 0)
(fdcy, fdcy_p)             = pF "fdcy" (Just 0)
(fsus, fsus_p)             = pF "fsus" (Just 0)
(frel, frel_p)             = pF "frel" (Just 0)


(vcavol, vcavol_p)         = pF "vcavol" (Just 0)
(outspread, outspread_p)   = pF "outspread" (Just 0)

(vamt, vamt_p)             = pF "vamt" (Just 0)
(vvel, vvel_p)             = pF "vvel" (Just 0)
(vdel, vdel_p)             = pF "vdel" (Just 0)

(lfo1rate, lfo1rate_p)     = pF "lfo1rate" (Just 0)
-- lfo1step                = makeF oscPolysynth "lfo1step"
(lfo1shape, lfo1shape_p)   = pF "lfo1shape" (Just 0)
(lfo1amt, lfo1amt_p)       = pF "lfo1amt" (Just 0)
(lfo1dest, lfo1dest_p)     = pF "lfo1dest" (Just 0)
(lfo1sync, lfo1sync_p)     = pF "lfo1sync" (Just 0)

(lfo2rate, lfo2rate_p)     = pF "lfo2rate" (Just 0)
-- lfo2step                = makeF oscPolysynth "lfo2step"
(lfo2shape, lfo2shape_p)   = pF "lfo2shape" (Just 0)
(lfo2amt, lfo2amt_p)       = pF "lfo2amt" (Just 0)
(lfo2dest, lfo2dest_p)     = pF "lfo2dest" (Just 0)
(lfo2sync, lfo2sync_p)     = pF "lfo2sync" (Just 0)

(lfo3rate, lfo3rate_p)     = pF "lfo3rate" (Just 0)
-- lfo3step                = makeF oscPolysynth "lfo3step"
(lfo3shape, lfo3shape_p)   = pF "lfo3shape" (Just 0)
(lfo3amt, lfo3amt_p)       = pF "lfo3amt" (Just 0)
(lfo3dest, lfo3dest_p)     = pF "lfo3dest" (Just 0)
(lfo3sync, lfo3sync_p)     = pF "lfo3sync" (Just 0)

(lfo4rate, lfo4rate_p)     = pF "lfo4rate" (Just 0)
-- lfo4step                = makeF oscPolysynth "lfo4step"
(lfo4shape, lfo4shape_p)   = pF "lfo4shape" (Just 0)
(lfo4amt, lfo4amt_p)       = pF "lfo4amt" (Just 0)
(lfo4dest, lfo4dest_p)     = pF "lfo4dest" (Just 0)
(lfo4sync, lfo4sync_p)     = pF "lfo4sync" (Just 0)

(emod, emod_p)             = pF "emod" (Just 0)
(eamt, eamt_p)             = pF "eamt" (Just 0)
(evel, evel_p)             = pF "evel" (Just 0)
(edel, edel_p)             = pF "edel" (Just 0)
(eatk, eatk_p)             = pF "eatk" (Just 0)
(edcy, edcy_p)             = pF "edcy" (Just 0)
(esus, esus_p)             = pF "esus" (Just 0)
(erel, erel_p)             = pF "erel" (Just 0)

(mod1src, mod1src_p)       = pF "mod1src" (Just 0)
(mod1amt, mod1amt_p)       = pF "mod1amt" (Just 0)
(mod1dst, mod1dst_p)       = pF "mod1dst" (Just 0)

(mod2src, mod2src_p)       = pF "mod2src" (Just 0)
(mod2amt, mod2amt_p)       = pF "mod2amt" (Just 0)
(mod2dst, mod2dst_p)       = pF "mod2dst" (Just 0)

(mod3src, mod3src_p)       = pF "mod3src" (Just 0)
(mod3amt, mod3amt_p)       = pF "mod3amt" (Just 0)
(mod3dst, mod3dst_p)       = pF "mod3dst" (Just 0)

(mod4src, mod4src_p)       = pF "mod4src" (Just 0)
(mod4amt, mod4amt_p)       = pF "mod4amt" (Just 0)
(mod4dst, mod4dst_p)       = pF "mod4dst" (Just 0)

(seq1dst, seq1dst_p)       = pF "seq1dst" (Just 0)
(seq2dst, seq2dst_p)       = pF "seq2dst" (Just 0)
(seq3dst, seq3dst_p)       = pF "seq3dst" (Just 0)
(seq4dst, seq4dst_p)       = pF "seq4dst" (Just 0)


(mwhl, mwhl_p)             = pF "mwhl" (Just 0)
(mwhldst, mwhldst_p)       = pF "mwhldst" (Just 0)

(aftt, aftt_p)             = pF "aftt" (Just 0)
(afttdst, afttdst_p)       = pF "afttdst" (Just 0)

(breath, breath_p)         = pF "breath" (Just 0)
(breathdst, breathdst_p)   = pF "breathdst" (Just 0)

(mvel, mvel_p)             = pF "mvel" (Just 0)
(mveldst, mveldst_p)       = pF "mveldst" (Just 0)

(foot, foot_p)             = pF "foot" (Just 0)
(footdst, footdst_p)       = pF "footdst" (Just 0)

(bendrng, bendrng_p)       = pF "bendrng" (Just 0)

-- left out: modwheel, breath, footctrl, pressure, velocity

(kbpm, kbpm_p)             = pF "kbpm" (Just 0)
(clockdiv, clockdiv_p)     = pF "clockdiv" (Just 0)

-- left out: pitchbend range

(sqntrig, sqntrig_p)       = pF "sqntrig" (Just 0)
(unisonkey, unisonkey_p)   = pF "unisonkey" (Just 0)
(unisonmode, unisonmode_p) = pF "unisonmode" (Just 0)
(arpmode, arpmode_p)       = pF "arpmode" (Just 0)

(erepeat, erepeat_p)       = pF "erepeat" (Just 0)
(unison, unison_p)         = pF "unison" (Just 0)


(arp, arp_p)               = pF "arp" (Just 0)
(sqn, sqn_p)               = pF "sqn" (Just 0)

(mcr1, mcr1_p)             = pF "mcr1" (Just 0)
(mcr2, mcr2_p)             = pF "mcr2" (Just 0)
(mcr3, mcr3_p)             = pF "mcr3" (Just 0)
(mcr4, mcr4_p)             = pF "mcr4" (Just 0)

(fbgain, fbgain_p)         = pF "fbgain" (Just 0)

(btnfreq, btnfreq_p)       = pF "btnfreq" (Just 0)
(btnvel, btnvel_p)         = pF "btnvel" (Just 0)
(btnmode, btnmode_p)       = pF "btnmode" (Just 0)

(sub1vol, sub1vol_p)       = pF "sub1vol" (Just 0)
(sub2vol, sub2vol_p)       = pF "sub2vol" (Just 0)

(fbvol, fbvol_p)           = pF "fbvol" (Just 0)

(ksplitpoint, ksplitpoint_p)     = pF "ksplitpoint" (Just 0)
(kmode, kmode_p)     = pF "kmode" (Just 0)
knormal         = kmode (p "0")
kstack          = kmode (p "0.5")
ksplit          = kmode (p "1")

-- sequences

seqtrack1 = map ((fst . (flip pF $ (Just 0))) . (\x -> "seq1step" ++ show x)) [1..16]
seqtrack2 = map ((fst . (flip pF $ (Just 0))) . (\x -> "seq2step" ++ show x)) [1..16]
seqtrack3 = map ((fst . (flip pF $ (Just 0))) . (\x -> "seq3step" ++ show x)) [1..16]
seqtrack4 = map ((fst . (flip pF $ (Just 0))) . (\x -> "seq4step" ++ show x)) [1..16]


seqpatterns p' = map onlyValues events
  where arc' = arc p'
        events = arc' (0,1)
        onlyValues = \(_, _, a) -> p a

sqnp i t p' = foldr (|+|) i (zipWith ($) t patterns)
  where patterns = seqpatterns p'

sqnp1 d p' = sqnp (seq1dst $ p $ show d) seqtrack1 p'
sqnp2 d p' = sqnp (seq2dst $ p $ show d) seqtrack2 p'
sqnp3 d p' = sqnp (seq3dst $ p $ show d) seqtrack3 p'
sqnp4 d p' = sqnp (seq4dst $ p $ show d) seqtrack4 p'


-- abstractions

-- adsr

atk' p' = attack p' |+| fatk p' |+| eatk p'
dcy' p' = decay p' |+| fdcy p' |+| edcy p'
sus' p' = sustain p' |+| fsus p' |+| esus p'
rel' p' = release p' |+| frel p' |+| erel p'

adsr a d s r = attack a |+| decay d |+| sustain s |+| release r
asr a d r = attack a |+| decay d |+| release r

-- lfo

lfotri = doublePattern 0
lforsaw = doublePattern 1
lfosaw = doublePattern 2
lfopulse = doublePattern 3
lforand = doublePattern 4

lrate r = ((min 150) . (max 0)) <$> p r
lstep s = ((min 166) . (max 151) . (+150)) <$> p s

lfo1 s d r a = lfo1shape s |+| lfo1dest d |+| lfo1rate r |+| lfo1amt a
lfo2 s d r a = lfo2shape s |+| lfo2dest d |+| lfo2rate r |+| lfo2amt a
lfo3 s d r a = lfo3shape s |+| lfo3dest d |+| lfo3rate r |+| lfo3amt a
lfo4 s d r a = lfo4shape s |+| lfo4dest d |+| lfo4rate r |+| lfo4amt a

--lfo = lfo1

-- mod

mod1 s d a = mod1src s |+| mod1dst d |+| mod1amt a
mod2 s d a = mod2src s |+| mod2dst d |+| mod2amt a
mod3 s d a = mod3src s |+| mod3dst d |+| mod3amt a
mod4 s d a = mod4src s |+| mod4dst d |+| mod4amt a

-- mod destination

doublePattern d = (p $ show d) :: Pattern Double

dosc1 = doublePattern 1
dosc2 = doublePattern 2
dosc = doublePattern 3
dmix = doublePattern 4
dnoise = doublePattern 5
dpw1 = doublePattern 6
dpw2 = doublePattern 7
dpw = doublePattern 8
dcut = doublePattern 9
dres = doublePattern 10
damod = doublePattern 11
dvca = doublePattern 12
dspread = doublePattern 13

dlfo1f = doublePattern 14
dlfo2f = doublePattern 15
dlfo3f = doublePattern 16
dlfo4f = doublePattern 17
dlfof = doublePattern 18

dlfo1a = doublePattern 19
dlfo2a = doublePattern 20
dlfo3a = doublePattern 21
dlfo4a = doublePattern 22
dlfoa = doublePattern 23

dfamt = doublePattern 24
dvamt = doublePattern 25
deamt = doublePattern 26
damt = doublePattern 27

dfatk = doublePattern 28
dvatk = doublePattern 29
deatk = doublePattern 30
datk = doublePattern 31

dfdcy = doublePattern 32
ddecay = doublePattern 33
dedcy = doublePattern 34
ddcy = doublePattern 35

dfrel = doublePattern 36
drelease = doublePattern 37
derel = doublePattern 38
drel = doublePattern 39

dmod1 = doublePattern 40
dmod2 = doublePattern 41
dmod3 = doublePattern 42
dmod4 = doublePattern 43

dfb = doublePattern 44
dsub1 = doublePattern 45
dsub2 = doublePattern 46
dfbgain = doublePattern 47
dslew = doublePattern 48

-- mod sources

sseq1 = doublePattern 1
sseq2 = doublePattern 2
sseq3 = doublePattern 3
sseq4 = doublePattern 4

slfo1 = doublePattern 5
slfo2 = doublePattern 6
slfo3 = doublePattern 7
slfo4 = doublePattern 8

sfenv = doublePattern 9
svenv = doublePattern 10
seenv = doublePattern 11

spitchb = doublePattern 12
smodwh = doublePattern 13
saftert = doublePattern 14
sbreath = doublePattern 15
sfoot = doublePattern 16
sexpr = doublePattern 17

svel = doublePattern 18
snote = doublePattern 19
snoise = doublePattern 20
-- drums


snare d p' = n p'
  |+| osc1shape zero |+| osc1kbd zero
  |+| osc2shape zero |+| osc2kbd zero
  |+| noise one
  |+| release d |+| sustain zero |+| decay d
  |+| cutoff (p "1")
  |+| dur (p "0.01")
  where zero = p "0"
        one = p "1"


kick blp p' = n p'
  |+| fourpole
  |+| osc1shape zero |+| osc1kbd zero
  |+| osc2shape zero |+| osc2kbd zero
  |+| sustain zero |+| decay (p "0.95") |+| release (p "0.5")
  |+| resonance (p "0.99") |+| cutoff zero
  |+| eamt (p "0.8") |+| emod (p "9") |+| edcy blp
  |+| dur (p "0.2")
  where zero = p "0"
